src/CoreLib/network.fs

```diff
@@ -170,55 +170,59 @@ type NetworkPerformance() =
         then
-            x.cmd <- _cmd
-            x.buffer <- _buffer
-            x.offset <- _offset
-            x.count <- _count
-    /// A new NetworkCommand formed from MemStream
-    /// 1. In case MemStream.Position == MemStream.Length, it is assumed
-    ///    the stream has just been written, offset is set to zero
-    /// 2. In case MemStream.Position <> MemStream.Length, it is assumed
-    ///    we are in middle of read, read continues from current position
-    new(_cmd, _ms : MemStream) as x =
+            x.cmd <- cmd
+            if (Utils.IsNotNull ms) then
+                x.ms <- ms
+                x.ms.AddRef()
+                // if position = length, assume stream just written, seek back for read purposes
+                if (ms.Position = ms.Length) then
+                    x.startPos <- 0L
+                else if (ms.Position <> 0L) then
+                    failwith "Use constructor with start position"
+            else
+                x.ms <- new MemStream(0)
+
+    new (cmd : ControllerCommand, ms : StreamBase<byte>, startPos : int64) as x =
         new NetworkCommand()
         then
-            x.cmd <- _cmd
-            if (Utils.IsNotNull _ms) then
-                x.buffer <- _ms.GetBuffer()
-                if (_ms.Position = _ms.Length) then
-                    // freshly written stream
-                    x.offset <- 0
-                    x.count <- int _ms.Length
-                else
-                    // stream read
-                    x.offset <- int _ms.Position
-                    x.count <- int (_ms.Length - _ms.Position)
+            x.cmd <- cmd
+            x.ms <- ms
+            x.ms.AddRef()
+            x.startPos <- startPos
+
+    member val startPos = 0L with get, set
     /// The controller command
     member val cmd = Unchecked.defaultof<ControllerCommand> with get, set
-    /// The buffer
-    member val buffer : byte[] = null with get, set
-    /// The buffer offset
-    member val offset : int = 0 with get, set
-    /// The buffer length (remaining)
-    member val count : int = 0 with get, set
-    member val internal buildHeader : bool = true with get, set
-    /// Get a MemStream object from the NetworkCommand
-    member x.MemStream() =
-        new MemStream(x.buffer, x.offset, x.count, false, true)
+    /// The memory stream
+    member val ms : StreamBase<byte> = null with get, set
     /// Get the length of the command on the wire
+    member val internal buildHeader : bool = true with get, set
     member x.CmdLen() =
-        if (x.buildHeader) then
-            8 + x.count
+        if (Utils.IsNull x.ms) then
+            if (x.buildHeader) then
+                8L
+            else
+                0L
         else
-            x.count
+            if (x.buildHeader) then
+                8L + x.ms.Length - x.startPos
+            else
+                x.ms.Length - x.startPos
+
+    /// release the underlying memstream
+    member x.Release() =
```

Comment:
I have added a IDisposable, however I have left Release in also.  It is a bit hard to wrap NetworkCommand in a using statement as it is in a queue, and when processing is done, it needs to be released.